### PR TITLE
Allow programatically making gadgets and querrying their names

### PR DESCRIPTION
**Pull request by RealKC**
This patch enables us to create a sidebar which would allow users to spawn whatever gadgets they choose.

### CODE REVIEW

File: gadgets/clock/src/lib.rs [hunk 1]
```diff
@@ -17,3 +13,19 @@ impl Gadget for ClockGadget {
         });
     }
 }
+
+pub struct ClockGadgetFactory;
```
Comment (lucastarche) [inline gadgets/clock/src/lib.rs]: You have the order `Gadget` `impl Gadget` `GadgetFactory` `impl GadgetFactory` here but `Gadget` `GadgetFactory` `impl Gadget` `impl GadgetFactory` for the other two.

Please choose one of the two and enforce it, so we get a consistent style across gadgets

**Review by lucastarche — COMMENTED** [review]
Overall looks good, please look into using one consistent style but besides that, looks fine

File: gadgets/clock/src/lib.rs (hunk 1 above)
Comment (RealKC) [reply gadgets/clock/src/lib.rs]: Ah that's a mistake, the style I'm trying to tend to is "struct definitions first, then impls, then free functions", will fix soon

File: src/main.rs [hunk 2]
```diff
@@ -1,26 +1,27 @@
...
+        for gadget_factory in gadgets::GADGET_FACTORIES {
+            gadgets.push(gadget_factory.make_gadget(&network_runtime, egui_ctx));
+        }
+
         Self {
-            background: AppBackground::default(),
-            clock_gadget: ClockGadget::new(&network_runtime, egui_ctx),
-            weather_gadget: WeatherGadget::new(&network_runtime, egui_ctx),
-            cses_status_gadget: CSESStatusGadget::new(&network_runtime, egui_ctx),
+            background: Default::default(),
```
Comment (lucastarche) [inline src/main.rs]: This reverts a change made in PR #7, use `AppBackground::default()` instead

Comment (RealKC) [reply src/main.rs]: Merges are my enemy lol

**Comment by lucastarche** [comment]
Looks good, thanks for the PR!

**Pull request merged**